File: 9-data_structures_stacks+queues/stack_ll.rb
# frozen_string_literal: true

class Node
  attr_accessor :next

  def initialize(value)
    @value = value
    @next = nil
  end
end

class Stack
  def initialize
    @top = nil
    @bottom = nil
    @length = 0
  end

  def peek
    @top
  end

  def push(value)
    new_node = Node.new(value)

    if @length.zero?
      @top = new_node
      @bottom = new_node
    else
      ref = @top
      @top = new_node
      @top.next = ref
    end

    @length += 1
  end

  def pop
    return nil if @top.nil?

    @bottom = nil if @top == @bottom

    @top = @top.next
    @length -= 1
  end
end

s = Stack.new
s.push('google')
s.push('udemy')
s.push('discord')
p s.peek
s.pop
s.pop
s.pop
p s

File: 9-data_structures_stacks+queues/stacks/lib/stack_ll.ex
defmodule Stack do
  defstruct lst: [], size: 0
  alias Stack, as: S

  def new(), do: %S{}

  def peek(%S{lst: []}), do: nil
  def peek(%S{lst: [h | _]}), do: h

  def push(lst, value) do
    %S{lst: [value: value, next: lst], size: lst.size + 1}
  end

  def pop(%S{lst: []}), do: nil
  def pop(s = %S{lst: [_ | t]}), do: %S{lst: t, size: s.size - 1}
end

# Stack.new()
# |> Stack.push('google')
# |> Stack.push('udemy')
# |> Stack.push('discord')
